### PR TITLE
GHA-79 Add UTs for PullRequestClosed

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
[GHA-79](https://sonarsource.atlassian.net/browse/GHA-79)

[GHA-79]: https://sonarsource.atlassian.net/browse/GHA-79?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
     "jest": "^29.7.0",
     "jest-editor-support": "^32.0.0-beta.1",
     "prettier": "^3.5.3",
-    "typescript": "^5.8.2"
+    "typescript": "^5.9.2"
```
Comment (pavel-mikula-sonarsource) [inline package.json:19]: Such a change should not be in this PR

File: tests/JiraClientStub.ts [hunk 2]
```diff
@@ -7,7 +7,8 @@ export const jiraClientStub = {
     switch (issueId) {
       case 'MMF-1111': return { key: 'MMF-1111', fields: { project: { key: 'MMF' }, issuetype: { name: 'Epic' } } };
       case 'KEY-1111': return { key: 'KEY-1111', fields: { project: { key: 'KEY' }, issuetype: { name: 'Epic' } } };
-      case 'KEY-1234': return { key: 'KEY-1234', fields: { project: { key: 'KEY' }, issuetype: { name: 'Task' } } };
+      case 'KEY-1234': return { key: 'KEY-1234', fields: { project: { key: 'KEY' }, issuetype: { name: 'Task' }, creator: {displayName: "CreatorKEY1234"} } };
```
Comment (pavel-mikula-sonarsource) [inline tests/JiraClientStub.ts:10]: It's "normal" user name, therefore with spaces
```suggestion
      case 'KEY-1234': return { key: 'KEY-1234', fields: { project: { key: 'KEY' }, issuetype: { name: 'Task' }, creator: {displayName: "Creator of KEY-1234"} } };
```

File: tests/JiraClientStub.ts [hunk 3]
```diff
@@ -41,6 +42,14 @@ export const jiraClientStub = {
       default: throw new Error(`Scaffolding did not expect team name ${accountId}`);
     }
   },
+  async findTransition(issueId: string, transitionName: string) {
+    switch (transitionName) {
+      case 'Merge into master': return { id: '10000', name: 'MergeMaster' };
+      case 'Merge into branch': return { id: '10001', name: 'MergeBranch' };
+     // case 'Resolve issue': return { id: '10002', name: 'MergeBranch' };
```
Comment (pavel-mikula-sonarsource) [inline tests/JiraClientStub.ts:49]: Unused comment, doesn't look intentional
```suggestion
```

File: tests/JiraClientStub.ts [hunk 4]
```diff
@@ -56,6 +65,9 @@ export const jiraClientStub = {
     fields = fields ? JSON.stringify(fields) : 'null';
     console.log(`Invoked jira.moveIssue('${issueId}', '${transitionName}', ${fields})`);
   },
+  async transitionIssue(issueId: string, transition: any, fields: any = null): Promise<void> {
+    console.log(`${issueId}: Executing '${transition.name}' (${transition.id}) transition`);
```
Comment (pavel-mikula-sonarsource) [inline tests/JiraClientStub.ts:69]: Use the same logging as every other `console.log`, to have the same format of `Invoked ... method name (serialized params)`

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
+import * as github from '@actions/github';
+import { PullRequestClosed } from './PullRequestClosed';
+import { LogTester } from '../tests/LogTester';
+import { jiraClientStub } from '../tests/JiraClientStub';
+import { createOctokitRestStub } from '../tests/OctokitRestStub';
+
+
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:7]:
```suggestion

```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+async function runAction(jiraProject: string, title: string, body?: string, user: string = 'test-user') {
+  process.env['INPUT_JIRA-PROJECT'] = jiraProject;
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:9]: It never changes => doesn't need to be parametrized
```suggestion
async function runAction(title: string, body?: string, user: string = 'test-user') {
  process.env['INPUT_JIRA-PROJECT'] = 'KEY';
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
+import * as github from '@actions/github';
+import { PullRequestClosed } from './PullRequestClosed';
+import { LogTester } from '../tests/LogTester';
+import { jiraClientStub } from '../tests/JiraClientStub';
+import { createOctokitRestStub } from '../tests/OctokitRestStub';
+
+
+async function runAction(jiraProject: string, title: string, body?: string, user: string = 'test-user') {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:8]: Never set => doesn't need to be here
```suggestion
async function runAction(jiraProject: string, title: string, user: string = 'test-user') {
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Invoked jira.moveIssue('KEY-5678', 'Cancel Issue', {\"resolution\":{\"id\":\"10001\"}})",
+      "Done"
+    ]);
+  });
+
+  it('PR merged on release branch', async () => {
+    github.context.payload.pull_request.merged = true;
+    github.context.payload.pull_request.base = { ref: 'master'};
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:92]: This needs a test against `master`, `main` and `feature/something`. foreach inside the UT would be fine. You'd need to clear the logTester params after each run. See #245

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  it('PR closed by Jira Tech User GitHub creator', async () => {
+    await runAction('KEY', 'KEY-5678 Title');
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Invoked jira.moveIssue('KEY-5678', 'Cancel Issue', {\"resolution\":{\"id\":\"10001\"}})",
+      "Done"
+    ]);
+  });
+
+  it('PR merged on release branch', async () => {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:90]:
```suggestion
  it('PR merged into release branch', async () => {
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    github.context.payload.pull_request.base = { ref: 'master'};
+    await runAction('KEY', 'KEY-1234 Title');
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "KEY-1234: Executing 'MergeMaster' (10000) transition",
+      "Done",
+    ]);
+  });
+
+  it('PR merged on normal branch', async () => {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:101]: I wasn't sure what "normal" means here
```suggestion
  it('PR merged into non-release branch', async () => {
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+  it('PR closed by Jira Tech User GitHub creator', async () => {
+    await runAction('KEY', 'KEY-5678 Title');
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Invoked jira.moveIssue('KEY-5678', 'Cancel Issue', {\"resolution\":{\"id\":\"10001\"}})",
+      "Done"
+    ]);
+  });
+
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:89]: `Merge into master` and `Merge into branch` are rare corner cases used by only a few projects. The main case for `Merge` transition is missing.

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    await runAction('KEY', 'Title', null, "renovate[bot]");
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Loading PR #42",
+      "Invoked jira.moveIssue('KEY-1234', 'Resolve issue', {\"resolution\":{\"id\":\"10001\"}})",
+      "Done",
+    ]);
+  });
+
+  it('PR closed by custom creator', async () => {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:72]: This is actually the case. See also the next comment to pair with this
```suggestion
  it('Unmerged PR for pre-existing issue closed', async () => {
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    await runAction('KEY', 'KEY-1234 Title');
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Skipping issue cancellation for creator CreatorKEY1234",
+      "Done"
+    ]);
+  });
+
+  it('PR closed by Jira Tech User GitHub creator', async () => {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:81]: This is the key part. Maybe put this test before `Unmerged PR for pre-existing issue closed` to make it more obvious
```suggestion
  it('Unmerged PR for ticket created by automation closed', async () => {
```

Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:81]: Test case is missing for case when the Jira issue is detected, but does not actually exist. Like `FAKE-123`

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    await runAction('KEY', 'KEY-1234 Title');
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Loading PR #42",
+      "Skipping issue resolution for non-Bot PR",
+      "Done"
+    ]);
+  });
+
+  it('is-eng-xp-squad Bot PR resolves issue', async () => {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:61]: This is what the UT does
```suggestion
  it('is-eng-xp-squad Bot PR unmerged PR cancels issue', async () => {
```

Comment (pavel-mikula-sonarsource) [reply PullRequestClosed/PullRequestClosed.test.ts:61]: The counter-example is missing: `is-eng-xp-squad Bot PR merged PR moves issue to Done`

**Review by pavel-mikula-sonarsource — APPROVED** [review]
Some testcases were not covered

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Apparently miss clicked :D

File: PullRequestClosed/PullRequestClosed.test.ts (hunk 11 above)
Comment (mary-georgiou-sonarsource) [reply PullRequestClosed/PullRequestClosed.test.ts:88]: How come? 
Is it due to having many non dev projects or something?

File: PullRequestClosed/PullRequestClosed.test.ts (hunk 13 above)
Comment (mary-georgiou-sonarsource) [reply PullRequestClosed/PullRequestClosed.test.ts:80]: I'm mostly testing the scaffolding here though - no?
added the test anyway.

File: tests/JiraClientStub.ts [hunk 15]
```diff
@@ -66,7 +66,7 @@ export const jiraClientStub = {
     console.log(`Invoked jira.moveIssue('${issueId}', '${transitionName}', ${fields})`);
   },
   async transitionIssue(issueId: string, transition: any, fields: any = null): Promise<void> {
-    console.log(`${issueId}: Executing '${transition.name}' (${transition.id}) transition`);
+    console.log(`Invoked jira.transitionIssue('${issueId}', 'transition: ${transition.name} ${transition.id}', ${fields}`);
```
Comment (pavel-mikula-sonarsource) [inline tests/JiraClientStub.ts:69]: If you send something else in that object, I want to see that
```suggestion
console.log(`Invoked jira.transitionIssue('${issueId}', ${JSON.stringify(transition)}, ${JSON.stringify(fields)}`);
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+import { PullRequestClosed } from './PullRequestClosed';
+import { LogTester } from '../tests/LogTester';
+import { jiraClientStub } from '../tests/JiraClientStub';
+import { createOctokitRestStub } from '../tests/OctokitRestStub';
+
+async function runAction(title: string, user: string = 'test-user') {
+  process.env['INPUT_JIRA-PROJECT'] = 'KEY';
+  const action = new PullRequestClosed();
+  (action as any).jira = jiraClientStub;
+  (action as any).rest = createOctokitRestStub(title, "", user);
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:12]: Same as the default value from the underlaying method
```suggestion
  (action as any).rest = createOctokitRestStub(title, null, user);
```

Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:12]: You can now use the trick from GHA-104
```suggestion
  const action = new PullRequestClosed() as PullRequestClosed & OctokitActionStub;
  action.jira = jiraClientStub;
  action.rest = createOctokitRestStub(title, "", user);
```

File: tests/JiraClientStub.ts [hunk 17]
```diff
@@ -56,6 +65,9 @@ export const jiraClientStub = {
     fields = fields ? JSON.stringify(fields) : 'null';
     console.log(`Invoked jira.moveIssue('${issueId}', '${transitionName}', ${fields})`);
   },
+  async transitionIssue(issueId: string, transition: any, fields: any = null): Promise<void> {
+    console.log(`Invoked jira.transitionIssue('${issueId}', 'transition: ${transition.name} ${transition.id}', ${fields}`);
```
Comment (pavel-mikula-sonarsource) [inline tests/JiraClientStub.ts:69]: Missing `)`
```suggestion
    console.log(`Invoked jira.transitionIssue('${issueId}', 'transition: ${transition.name} ${transition.id}', ${fields})`);
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+
+  it('PR merged into release branch', async () => {
+    const releaseBranches = ['master', 'main', 'branch-0.0.0'];
+    for (const branchName of releaseBranches) {
+      github.context.payload.pull_request.merged = true;
+      github.context.payload.pull_request.base = { ref: branchName };
+      await runAction('KEY-1234 Title');
+      expect(logTester.logsParams).toStrictEqual([
+        "Loading PR #42",
+        `Invoked jira.transitionIssue('KEY-1234', 'transition: Merge into master 10000', null`,
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:121]: This can have `"` instead of backtick once it fails after updating the stub. Just copy/paste what the error fails with, including the ugly backlashes

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+    github.context.payload.pull_request.base = { ref: 'user/branch' };
+    await runAction('KEY-5678 Title');
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Invoked jira.transitionIssue('KEY-5678', 'transition: Merge into branch 10001', null",
+      "Done",
+    ]);
+  });
+
+  it('Merge Pull Request - transision is missing', async () => {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:139]:
```suggestion
  it('Merge PR for workflow without feature branches', async () => {
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    github.context.payload.pull_request.merged = true;
+    github.context.payload.pull_request.base = { ref: 'user/branch' };
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:141]: This shouldn't matter, let's use the usual suspect
```suggestion
    github.context.payload.pull_request.base = { ref: 'master' };
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+      "Skipping issue cancellation for creator Creator of KEY-1234",
+      "Done"
+    ]);
+  });
+
+  it('Jira issue does not exist', async () => {
+    const logSpy = jest.spyOn(core, 'setFailed').mockImplementation(() => { });
+    try {
+      await runAction('FAKE-1234 Title');
+      expect(logSpy).toHaveBeenCalledWith('Scaffolding did not expect FAKE-1234');
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:107]: Nope, you should have updated the scaffolding to case `FAKE-1234` and return `null`. The current version doesn't test the action, it tests the scaffolding

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+    await runAction('Title', "renovate[bot]");
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Loading PR #42",
+      "Invoked jira.moveIssue('KEY-1234', 'Resolve issue', {\"resolution\":{\"id\":\"10001\"}})",
+      "Done",
+    ]);
+  });
+
+  it('Unmerged PR for ticket created by automation closed', async () => {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:85]:
```suggestion
  it('Unmerged PR for ticket created by automation is closed', async () => {
```

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+    await runAction('KEY-5678 Title');
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Invoked jira.moveIssue('KEY-5678', 'Cancel Issue', {\"resolution\":{\"id\":\"10001\"}})",
+      "Done"
+    ]);
+  });
+
+  it('Unmerged PR for pre-existing issue closed', async () => {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:94]:
```suggestion
  it('Unmerged PR for pre-existing issue is closed', async () => {
```

File: PullRequestClosed/PullRequestClosed.test.ts (hunk 13 above)
Comment (pavel-mikula-sonarsource) [reply PullRequestClosed/PullRequestClosed.test.ts]: Well...

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
One more round

File: PullRequestClosed/PullRequestClosed.test.ts (hunk 18 above)
Comment (mary-georgiou-sonarsource) [reply PullRequestClosed/PullRequestClosed.test.ts]: Actually I had a parameter in there and removed it - so the backticks are a remainder.

File: PullRequestClosed/PullRequestClosed.test.ts [hunk 24]
```diff
@@ -101,13 +101,12 @@ describe('PullRequestClosed', () => {
...
-    try {
-      await runAction('FAKE-1234 Title');
-      expect(logSpy).toHaveBeenCalledWith('Scaffolding did not expect FAKE-1234');
-    } finally {
-      logSpy.mockRestore();
-    }
+    await runAction('FAKE-1234 Title');
+    expect(logTester.logsParams).toStrictEqual([
+      "Loading PR #42",
+      "Skipping issue cancellation for creator undefined",
```
Comment (pavel-mikula-sonarsource) [inline PullRequestClosed/PullRequestClosed.test.ts:107]: Please add `|| null` in `PullRequestClosed` line 31 to have `creator null` here. I was fixing these surprises when I discovered them

Comment (mary-georgiou-sonarsource) [reply PullRequestClosed/PullRequestClosed.test.ts]: Illegal

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM
